File: src/process_image.ts
import linalg from "./linalg"

const get_face_bounding_box = (face: number[], vertices: number[][]): number[] =>{
    const v1 = vertices[face[0]];
    const v2 = vertices[face[1]];
    const v3 = vertices[face[2]];

    return [
        Math.min(v1[0], v2[0], v3[0]),
        Math.max(v1[0], v2[0], v3[0]),
        Math.min(v1[1], v2[1], v3[1]),
        Math.max(v1[1], v2[1], v3[1]),
    ].map(val => Math.round(val));
}

const get_pixel_depth = (x: number, y: number, face: number[], vertices: number[][]): number =>{
    const v1 = vertices[face[0]];
    const v2 = vertices[face[1]];
    const v3 = vertices[face[2]];

    const dir1 = linalg.vecsubtract(v2,v1);
    const dir2 = linalg.vecsubtract(v3,v1);

    const norm = linalg.veccross(dir1,dir2);
    return (linalg.vecdot(norm,v1) - norm[0]*x - norm[1]*y) / norm[2];
}

const pixel_face_collision = (x: number, y: number, face: number[], vertices: number[][]): boolean =>{
    let v1 = vertices[face[0]];
    let v2 = vertices[face[1]];
    let v3 = vertices[face[2]];
    let p = [x,y,0];
    

    // Recenter so v1 is at origin
    v2 = linalg.vecsubtract(v2,v1);
    v3 = linalg.vecsubtract(v3,v1);
    p = linalg.vecsubtract(p,v1);

    const mat = [
        [v2[0],v3[0]],
        [v2[1],v3[1]]
    ];

    const inv = linalg.matinv(mat);
    if(inv === false) return false;

    const weights = linalg.matvec(inv,[p[0],p[1]]);
    const w1 = weights[0];
    const w2 = weights[1];
    return 0 <= w1 && w1 <= 1 &&
        0 <= w2 && w2 <= 1 &&
        0 <= w2+w2 && w1+w2 <= 1;
}

const computeLightIntensity = (face: number[], vertices: number[][], lightPos: number[]): number =>{
    const v1 = vertices[face[0]];
    const v2 = vertices[face[1]];
    const v3 = vertices[face[2]];

    const dir1 = linalg.vecsubtract(v2,v1);
    const dir2 = linalg.vecsubtract(v3,v1);
    const normal = linalg.vecnormalize(linalg.veccross(dir1,dir2));
    const faceToLightSrc = linalg.vecnormalize(linalg.vecsubtract(lightPos, v1));

    return Math.abs(linalg.vecdot(normal, faceToLightSrc));
}

const make_pixelbuffer = (height: number, width: number, faces: number[][], vertices: number[][], objcolor: number[]): number[][][] =>{
    let pixelbuffer = linalg.makemat(height, width, [0,0,0,255]);
    let zbuffer = linalg.makemat(height, width, Number.MIN_SAFE_INTEGER);
    const lightPos = [0, 0, Number.MAX_SAFE_INTEGER];

    faces.forEach(face =>{
        const face_bounds = get_face_bounding_box(face, vertices);
        for(let row = Math.max(0, face_bounds[2]); row <= Math.min(height-1, face_bounds[3]); ++row){
            for(let col = Math.max(0, face_bounds[0]); col <= Math.min(width-1, face_bounds[1]); ++col){
                if(pixel_face_collision(col,row,face,vertices)){
                    const depth = get_pixel_depth(col,row,face,vertices);
                    if(depth >= zbuffer[row][col]){
                        zbuffer[row][col] = depth;
                        //Diffuse shading
                        const intensity = computeLightIntensity(face, vertices, lightPos);
                        pixelbuffer[row][col] = objcolor.map(x => x*intensity);
                    }
                }
            }
        }
    });
    return pixelbuffer;
}

export default{
    make_pixelbuffer
}

File: process_image.py
from typing import List
import numpy as np
import math

def pixel_depth(x: int, y: int, face: List[int], vertices: List[List[float]]) -> float:
    v1 = vertices[face[0]]
    v2 = vertices[face[1]]
    v3 = vertices[face[2]]

    dir1 = np.subtract(v2,v1)
    dir2 = np.subtract(v3,v1)

    normal: List[float] = np.cross(dir1,dir2)
    z = (np.dot(normal, v1) - (normal[0]*x) - (normal[1]*y)) / normal[2]
    return z

def get_face_bounding_box(face: List[int], vertices: List[List[float]]) -> List[int]:
    v1 = vertices[face[0]]
    v2 = vertices[face[1]]
    v3 = vertices[face[2]]
    v_list = [v1,v2,v3]

    x_coords = list(map(lambda v: v[0], v_list))
    y_coords = list(map(lambda v: v[1], v_list))
    return list(map(lambda n: round(n), 
        [min(x_coords), max(x_coords), min(y_coords), max(y_coords)]))

def is_pixel_on_face(x: int, y: int, face: List[int], vertices: List[List[float]]) -> bool:
    v1 = vertices[face[0]]
    v2 = vertices[face[1]]
    v3 = vertices[face[2]]
    p = [x,y,0]

    # recenter so v1 is at origin
    v2 = np.subtract(v2, v1)
    v3 = np.subtract(v3, v1)
    p = np.subtract(p, v1)

    denom = np.linalg.det([
        [v2[0], v3[0]],
        [v2[1], v3[1]]
    ])

    if denom == 0:
        return False
    
    w2 = (p[0]*v3[1] - p[1]*v3[0]) / denom
    w3 = (p[1]*v2[0] - p[0]*v2[1])  / denom

    return (0 <= w2 and w2 <= 1 and 
            0 <= w3 and w3 <= 1 and 
            0 <= w2 + w3 and w2 + w3 <= 1)

def make_pixelbuffer(height: int, width: int, faces: List[List[int]], vertices: List[List[float]], colors: List[List[int]]) -> List[List[List[int]]]:
    pixelbuffer: List[List[List[int]]] = np.zeros((height, width, 3), dtype='uint8') 
    zbuffer: List[List[float]] = np.full((height, width), -math.inf, dtype=np.float32)

    for i, face in enumerate(faces):
        print(f'Rendering face {i+1}/{len(faces)}')
        face_bounding_box = get_face_bounding_box(face, vertices)

        # for row in range(face_bounding_box[2], face_bounding_box[3] + 1):
        #     for col in range(face_bounding_box[0], face_bounding_box[1] + 1):
        #         if is_pixel_on_face(col, row, face, vertices):
        #             if row >= height or col >= width or row < 0 or col < 0: continue
                    
        #             depth = pixel_depth(col, row, face, vertices)
        #             if depth >= zbuffer[row,col]:
        #                 zbuffer[row,col] = depth
        #                 pixelbuffer[row,col] = colors[i]
        
        row = max(face_bounding_box[2],0)
        while row <= min(face_bounding_box[3], height-1):
            col = max(face_bounding_box[0], 0)
            while col <= min(face_bounding_box[1], width-1):
                if is_pixel_on_face(col, row, face, vertices):
                    depth = pixel_depth(col, row, face, vertices)
                    if depth >= zbuffer[row,col]:
                        zbuffer[row,col] = depth
                        pixelbuffer[row,col] = colors[i]
                col += 1
            row += 1

    return pixelbuffer